tlair/src/main/scala/at/forsyte/apalache/tla/lir/TypeLevelBuilder.scala

```diff
@@ -0,0 +1,186 @@
+  }
+}
+
+// These examples illustrate construction of expressions, and, commented out,
+// constructions which are impossible at compile time, due to invalid construction
+object Example {
+  import TypeLevelBuilder._
+
+  // Atomic values
+  val exInt: Tla.Int = Tla.Int(1)
+  val exStr: Tla.Str = Tla.Str("foo")
+  val exFalse: Tla.Bool = Tla.Bool(false)
+  val exTrue: Tla.Bool = Tla.Bool(true)
+
+  val exPlus = plus(exInt, exInt)
+
+  // NOTE: Statically fails with type error
+  // val exInvalidPlus = plus(exInt, exStr)
+  // Will statically fail with:
+  // [E]      type mismatch;
+  // [E]       found   : at.forsyte.apalache.tla.lir.TypeLevelTlaEx.Tla.Str
+  // [E]       required: at.forsyte.apalache.tla.lir.TypeLevelTlaEx.Tla.Int
+  // [E]      L116:     val exBadPlus = Builder.plus(exInt, exStr)
+  // [E]                                                    ^^^^^
+
+  // Emonstration of polymorphic use of `equal`
+  val exIntEquals = equal(exInt, exInt)
+  val exStrEquals = equal(exStr, exStr)
+
+  // NOTE: Statically fails with type error
+  // val exInvalidEquals = equal(exStr, exInt)
+
+  val exBinaryAnd = and(exIntEquals, exStrEquals)
+  val exVariadicAnd = and(exBinaryAnd, Tla.Bool(true), Tla.Bool(false), Tla.Bool(false))
+
+  // NOTE: Statically fails with type error
+  // val exInvalidAnd = and(exInt, Tla.Bool(true))
+
+  val exIntSeq = seq(Tla.Int(1), Tla.Int(2), Tla.Int(3))
+  val exStrSeq = seq(Tla.Str("one"), Tla.Str("two"), Tla.Str("three"))
+
+  // Type inference will give us differently typed empty seqs
+  val exEmptyIntSeq: Tla.TlaSeq[Tla.IntT] = seq()
+  val exEmptyStrSeq: Tla.TlaSeq[Tla.Str] = seq()
+  val exEmptyBoolSeq = seq[Tla.Bool]()
+
+  // NOTE: Statically fails with type error
+  // val exInvalidSeq = seq(Tla.Str("one"), Tla.Int(2))
+
+  val exTuple: Tla.Tup[Tla.Int, Tla.Str] = tup(exInt, exStr)
+  val exNestedTuple = tup(exTrue, exTuple)
+  val exTupleOfSeq = tup(seq(Tla.Int(1), Tla.Int(2)), exStr)
+  val exSeqOfTuple = seq(tup(exInt, exStr), tup(Tla.Int(1), Tla.Str("bar")))
+
+  // NOTE: Statically fails with type error, due to clash with annotation
+  // val exInvalidTuple: Tla.Tup[Tla.Str, Tla.Str] = tup(Tla.Str("one"), Tla.Int(2))
+  // NOTE: Statically fails with type error, due to incompatible tuples
+  // val exInvalidSeqOfTuple = seq(tup(exInt, exStr), tup(exStr, exInt))
+
+  // We need some extra machinery for records, but Shapeless supports type-level key-value
+  // encoding
+  import shapeless._
+  import record._
+  import syntax.singleton._
+
+  val exRecord = Tla.Rec(("msg" ->> Tla.Str("hello")) :: ("value" ->> Tla.Int(3)) :: HNil)
+  val exRecordAccess1 = plus(exRecord.fields("value"), Tla.Int(3))
+  val exRecordAccess2 = equal(Tla.Str("hello"), exRecord.fields("msg"))
+
+  // NOTE: Statically fails with type error, due to invalid record field type
+  // I.e., field "msg" holds a string, so we cannot use it's value in `plus`.
+  // val exInvalidRecordAccess = plus(exRecord.fields("msg"), Tla.Int(3))
```

Comment:
Wow. That's magic! Do you know how it works?

Reply:
The constructor generates a singleton field-types for the label, that encode the value as a parameter! :D

https://github.com/milessabin/shapeless/blob/36089a7c0eb009dd9dc313bed44df822fea8a96e/core/src/main/scala/shapeless/labelled.scala#L80-L85